### PR TITLE
Fixed a crash when refineries share a cell

### LINKED ISSUE #8467: Crash in the harvester logic
Happened while testing soviet-05 (see also #8446).

```
Red Alert Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.ArgumentException`: An item with the same key has already been added. (Ein Element mit dem gleichen Schlüssel wurde bereits hinzugefügt.)
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at OpenRA.Mods.Common.Traits.Harvester.ClosestProc(Actor self, Actor ignore)
   at OpenRA.Mods.Common.Traits.Harvester.ChooseNewProc(Actor self, Actor ignore)
   at OpenRA.Mods.Common.Traits.Harvester.<>c__DisplayClass4.<.ctor>b__2()
   at OpenRA.Activities.CallFunc.Tick(Actor self) in \OpenRA\OpenRA.Game\Activities\CallFunc.cs:Line 29.
   at OpenRA.Traits.Util.RunActivity(Actor self, Activity act) in \OpenRA\OpenRA.Game\Traits\Util.cs:Line 83.
   at OpenRA.Actor.Tick() in \OpenRA\OpenRA.Game\Actor.cs:Line 121.
   at OpenRA.World.Tick() in \OpenRA\OpenRA.Game\World.cs:Line 302.
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in \OpenRA\OpenRA.Game\Game.cs:Line 456.
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:Line 480.
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:Line 596.
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:Line 636.
   at OpenRA.Program.Run(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 111.
   at OpenRA.Program.Main(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 39.

```
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/21749489-crash-in-the-harvester-logic?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Mailaender**
Fixes https://github.com/OpenRA/OpenRA/issues/8467.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 1]
```diff
@@ -204,7 +205,11 @@ public Actor ClosestProc(Actor self, Actor ignore)
 				path = self.World.WorldActor.Trait<IPathFinder>().FindPath(search);
 
 			if (path.Count != 0)
-				return refs[path.Last()].Actor;
+			{
+				var closestRefinery = refs.FirstOrDefault(r => r.Location == path.Last());
+				if (closestRefinery != null)
```
Comment (tovl) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: We could use the new syntax here:
```C#
return refs.FirstOrDefault(r => r.Location == path.Last())?.Actor;
```

**Comment by tovl** [comment]
A naive search is less performance-friendly than the hash-table lookup this replaces, but that shouldn't matter considering the number of items would typically be in the single digits anyway. Code changes looks good, but I haven't tested in-game yet.

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 2]
```diff
@@ -181,18 +181,19 @@ public Actor ClosestProc(Actor self, Actor ignore)
 					Location = r.Actor.Location + r.Trait.DeliveryOffset,
 					Actor = r.Actor,
 					Occupancy = self.World.ActorsHavingTrait<Harvester>(h => h.LinkedProc == r.Actor).Count()
-				}).ToDictionary(r => r.Location);
+				}).ToList();
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: Think we could use ToLookup here to have the best of both worlds.

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/Harvester.cs]: That sounds smart.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b99caf03a460df049351234a8dd41d4a33386eb6...a93b412eb2605aef89792627f175725d242fc43b)

**Pull request merged**